repos/keg-core/core/base/hooks/state/useExternalSet.js

```diff
@@ -0,0 +1,126 @@
+import { useCallback, useMemo } from 'react'
+import { uniqArr, identity, omitElement, pipeline } from '@keg-hub/jsutils'
+
+/**
+ * Helper for `useExternalSet`, checking if element
+ * is a member of `arr`, using the `selector` to
+ * get the right comparison value.
+ * @param {Array<*>} arr 
+ * @param {*} element 
+ * @param {Function?} selector - gets the prop to compare the elements with
+ */
+const isMember = (arr, element, selector=identity) => {
+  const id = selector(element)
+  return arr?.some(
+    item => selector(item) === id
+  )
+}
+
+/**
+ * Provides a set-like interface to an external array `arr`.
+ * If it determines a mutation needs to occur, it will call `setArr` with
+ * the a new array that has the mutation applied.
+ * 
+ * By default, useExternalSet determines if an element exists in the set
+ * using reference equality. If you would like to use a different method,
+ * pass in your own comparison function.
+ * 
+ * 
+ * @param {Array} arr
+ * @param {Function} setArr
+ * @param {Function?} selector - optional fn to used to get the property for checking element membership. 
+ *  Must have signature: (elementToCheck, existingElement) => boolean (is equal)
+ *  Usually only necessary for non-primitive types.
+ * @returns {Object} set-like interface to arr
+ * 
+ * @example
+ * const users = useSelector(store => store.items.userIds) 
+ * const usersSet = useExternalSet(users, setUsersInStore, (a, b) => a.id === b.id)
+ * usersSet.data // === [ { id: '1' }, { id: '2'}, { id: '3'} ]
+ * usersSet.delete({ id: '4'}) // does nothing, since the id doesn't exist
+ * usersSet.add({ id: '4' }) // calls setUsersInStore with an array including additional user with id 4
+ * usersSet.has({ id: '4' }) // true
+ * usersSet.add({ id: '4' }) // does nothing
+ * usersSet.delete({ id: '4' }) // calls setUsersInStore with an array that omits user id 4
+ */
+export const useExternalSet = (arr, setArr, selector=identity) => {
+  // checks if arr contains the element, using the comparisonFn if defined
+  const contains = useCallback(
+    (arr, element) => isMember(arr, element, selector),
+    [ selector ],
+  )
+
+  return useMemo(
```

Comment:
This is dope but why define it inline?
* Every time the hook is called, all these set methods have to be recreated.
* It seems like you could define it outside of the useMemo hook, and then reuse it?

But maybe I'm missing something

Reply:
Because they each depend on the `arr` argument, and so if they were defined separately, they would either get out of date quickly or would need to be recreated separately.

I _did_ consider using a class, something like:

```js
class ExternalSet {
  constructor(array, setArray) {
    this.array = array
    this.setArrray = setArray
  }

  add (element) {
    pipeline([ ...arr, element ], uniqArr, setArr))
  }
 
   ...
}
```

and the hook would be something like:

```js
const useExternalSet = (arr, setArr, selector) => {
  return useMemo(() => new ExternalSet(arr, setArr, selector), [ arr, setArr, selector ])
}
```

Of course, now it needs to instantiate a new `ExternalSet` class instance every time instead of just an object literal with functions. I'm not sure that would be any faster though (and historically I've seen classes be really slow). On the other hand, it might not recreate the functions, since it could just reuse the ones defined on the prototype.

In the end, I thought any performance differences with this would be minor, but using the class would be more complex, so I chose the object instead.

Reply:
wow, nice explanation. Seems like you really spent time thinking about it. That works for me.